pkg/cortex/serve/cortex_internal/lib/api/task.py

```diff
@@ -92,38 +129,10 @@ def _read_impl(self, module_name: str, impl_path: str, target_class_name: str):
         else:
             return callables[0]
 
-    def _validate_impl(self, impl):
+    @staticmethod
+    def _validate_impl(impl):
         if inspect.isclass(impl):
-            target_class_name = impl.__name__
-
-            constructor_fn = getattr(impl, "__init__", None)
-            if constructor_fn:
-                argspec = inspect.getfullargspec(constructor_fn)
-                if not (len(argspec.args) == 1 and argspec.args[0] == "self"):
-                    raise UserException(
-                        f"class {target_class_name}",
-                        f'invalid signature for method "__init__"',
-                        f'only "self" parameter must be present in method\'s signature',
-                    )
-
-            callable_fn = getattr(impl, "__call__", None)
-            if callable_fn:
-                argspec = inspect.getfullargspec(callable_fn)
-                if not (
-                    len(argspec.args) == 2
-                    and argspec.args[0] == "self"
-                    and argspec.args[1] == "config"
-                ):
-                    raise UserException(
-                        f"class {target_class_name}",
-                        f'invalid signature for method "__call__"',
-                        f'the following parameters must be present in method\'s signature: "self", "config"',
-                    )
-            else:
-                raise UserException(
-                    f"class {target_class_name}",
-                    f'"__call__" method not defined',
-                )
+            validate_class_impl(impl, TASK_CLASS_VALIDATION)
         else:
             callable_fn = impl
             argspec = inspect.getfullargspec(callable_fn)
```

Comment:
If the user provides a callable (i.e. function), then they wouldn't be able to use the metrics client?

Reply:
I opted for that approach for consistency with the rest of the APIs.

The class version becomes the "advanced use case" while the function becomes the straightforward one